sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -927,6 +942,185 @@ abstract class GpuBaseAggregateMeta[INPUT <: SparkPlan](
+        val aggMeta = aggregateExpressions(i).childExprs.head
+            .asInstanceOf[TypedImperativeAggExprMeta[_]]
+        val desiredDataType = aggMeta.aggBufferAttribute.dataType
+
+        var buf = aggFn.aggBufferAttributes.head
+        desiredAggBufTypes(buf.exprId) = desiredDataType
+
+        buf = aggFn.inputAggBufferAttributes.head
+        desiredInputAggBufTypes(buf.exprId) = desiredDataType
+
+      case _ =>
+    }
+
+    // Overrides the data types of typed imperative aggregation buffers for type checking
+    aggregateAttributes.foreach { attrMeta =>
+      attrMeta.wrapped match {
+        case ar: AttributeReference if desiredAggBufTypes.contains(ar.exprId) =>
+          attrMeta.overrideDataType(desiredAggBufTypes(ar.exprId))
+        case _ =>
+      }
+    }
+    resultExpressions.foreach { retMeta =>
+      retMeta.wrapped match {
+        case ar: AttributeReference if desiredInputAggBufTypes.contains(ar.exprId) =>
+          retMeta.overrideDataType(desiredInputAggBufTypes(ar.exprId))
+        case _ =>
+      }
+    }
+  }
+}
+
+object GpuTypedImperativeSupportedAggregateExecMeta {
+
+  private val entireAggFallbackCheck = TreeNodeTag[Boolean](
+    "rapids.gpu.checkAndFallbackAggregateExecEntirely")
+
+  private def checkAndFallbackEntirely(
+      meta: GpuTypedImperativeSupportedAggregateExecMeta[_]): Unit = {
+    // We only run the check for final stages which contain TypedImperativeAggregate.
+    val needToCheck = meta.agg.aggregateExpressions.exists(e =>
+      (e.mode == Final || e.mode == Complete) &&
+          e.aggregateFunction.isInstanceOf[TypedImperativeAggregate[_]])
+    if (!needToCheck) return
+    // Avoid duplicated check and fallback.
+    val checked = meta.agg.getTagValue[Boolean](entireAggFallbackCheck).contains(true)
+    if (checked) return
+
+    meta.agg.setTagValue(entireAggFallbackCheck, true)
+    val logicalPlan = meta.agg.logicalLink.get
+    val stageMetas = mutable.ListBuffer[GpuBaseAggregateMeta[_]]()
+    // Go through all Aggregate stages to check whether all stages is GPU supported. If not,
+    // we fall back all GPU supported stages to CPU.
+    if (recursiveCheckForFallback(meta, logicalPlan, stageMetas)) {
+      stageMetas.foreach {
+        case aggMeta if aggMeta.canThisBeReplaced =>
+          aggMeta.willNotWorkOnGpu("Associated fallback for TypedImperativeAggregate")
+        case _ =>
+      }
+    }
+  }
+
+  /**
+   * Recursively collect all PlanMetas of input LogicalPlan. At the same time, and check whether
+   * existing plans which can NOT be replaced among them. If any, return true as the label of
+   * the entire fallback.
+   */
+  private def recursiveCheckForFallback(
+      currentMeta: SparkPlanMeta[_],
+      logical: LogicalPlan,
+      metaOfAllStages: mutable.ListBuffer[GpuBaseAggregateMeta[_]]): Boolean = {
+    currentMeta match {
+      case aggMeta: GpuBaseAggregateMeta[_] if aggMeta.agg.logicalLink.contains(logical) =>
```

Comment:
This looks like it searches the entire query plan, but just looking at the first part of the query.  If any of them cannot be replaced, then none of them will be replaced. So what happens if there are multiple aggregations that take place in a single query? It looks like we are now at a point where even one aggregation that falls back to the CPU will cause all of them to fall back after it. Even if they are not TypeImperative aggregates.

Reply:
I believe it will only fall back AggregateExecs created from same logical plan as the target stage. I added a test to verify that no other aggregations will be affected if there exists more than one Aggregation plans: `test_hash_groupby_collect_partial_replace_fallback_with_other_agg`.